### PR TITLE
Convert  hashing, partitioning, and nested_loop_join APIs to use device_uvector

### PR DESCRIPTION
**Pull request by harrism**
Converts internals of cudf::partition/hash_partition, hashing, and nested_loop_join to use device_uvector instead of device_vector.

Changes `cudf::hash` API to take a `host_span` instead of a `std::vector`, hence marked as a breaking change.

Contributes to #7287

Performance is improved in hash_partition for smaller sizes, and is similar for larger sizes.

```
(rapids) rapids@compose:~/cudf/cpp/build/release$ _deps/benchmark-src/tools/compare.py benchmarks ~/cudf/cpp/build/hashing_before.json ~/cudf/cpp/build/hashing_after.json
Comparing /home/mharris/rapids/cudf/cpp/build/hashing_before.json to /home/mharris/rapids/cudf/cpp/build/hashing_after.json
Benchmark                                                             Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------
Hashing/hash_partition/131072/1/64/manual_time                     -0.2817         -0.2422             0             0             0             0
Hashing/hash_partition/262144/1/64/manual_time                     -0.2561         -0.2208             0             0             0             0
Hashing/hash_partition/524288/1/64/manual_time                     -0.2269         -0.1992             0             0             0             0
Hashing/hash_partition/1048576/1/64/manual_time                    -0.1864         -0.1683             0             0             0             0
Hashing/hash_partition/2097152/1/64/manual_time                    -0.1394         -0.1323             0             0             0             0
Hashing/hash_partition/131072/1/128/manual_time                    -0.2732         -0.2340             0             0             0             0
Hashing/hash_partition/262144/1/128/manual_time                    -0.2561         -0.2218             0             0             0             0
Hashing/hash_partition/524288/1/128/manual_time                    -0.2223         -0.1946             0             0             0             0
Hashing/hash_partition/1048576/1/128/manual_time                   -0.1847         -0.1664             0             0             0             0
Hashing/hash_partition/2097152/1/128/manual_time                   -0.1345         -0.1260             0             0             0             0
Hashing/hash_partition/131072/1/256/manual_time                    -0.2626         -0.2280             0             0             0             0
Hashing/hash_partition/262144/1/256/manual_time                    -0.2545         -0.2206             0             0             0             0
Hashing/hash_partition/524288/1/256/manual_time                    -0.2178         -0.1918             0             0             0             0
Hashing/hash_partition/1048576/1/256/manual_time                   -0.1739         -0.1577             0             0             0             0
Hashing/hash_partition/2097152/1/256/manual_time                   -0.1389         -0.1316             0             0             0             0
Hashing/hash_partition/131072/1/512/manual_time                    -0.2492         -0.2171             0             0             0             0
Hashing/hash_partition/262144/1/512/manual_time                    -0.2383         -0.2060             0             0             0             0
Hashing/hash_partition/524288/1/512/manual_time                    -0.1971         -0.1758             0             0             0             0
Hashing/hash_partition/1048576/1/512/manual_time                   -0.1679         -0.1529             0             0             0             0
Hashing/hash_partition/2097152/1/512/manual_time                   -0.1426         -0.1394             0             0             0             0
Hashing/hash_partition/131072/1/1024/manual_time                   -0.2246         -0.1952             0             0             0             0
Hashing/hash_partition/262144/1/1024/manual_time                   -0.2153         -0.1880             0             0             0             0
Hashing/hash_partition/524288/1/1024/manual_time                   -0.1851         -0.1664             0             0             0             0
Hashing/hash_partition/1048576/1/1024/manual_time                  -0.1456         -0.1350             0             0             0             0
Hashing/hash_partition/2097152/1/1024/manual_time                  -0.1190         -0.1141             0             0             0             0
Hashing/hash_partition/131072/16/64/manual_time                    -0.0639         -0.0638             0             0             0             0
Hashing/hash_partition/262144/16/64/manual_time                    -0.0462         -0.0450             0             0             0             0
Hashing/hash_partition/524288/16/64/manual_time                    -0.0354         -0.0368             1             1             1             1
Hashing/hash_partition/1048576/16/64/manual_time                   -0.0152         -0.0182             1             1             1             1
Hashing/hash_partition/2097152/16/64/manual_time                   -0.0142         -0.0151             2             2             2             2
Hashing/hash_partition/131072/16/128/manual_time                   -0.0618         -0.0604             0             0             0             0
Hashing/hash_partition/262144/16/128/manual_time                   -0.0472         -0.0463             0             0             0             0
Hashing/hash_partition/524288/16/128/manual_time                   -0.0383         -0.0383             1             1             1             1
Hashing/hash_partition/1048576/16/128/manual_time                  -0.0172         -0.0173             1             1             1             1
Hashing/hash_partition/2097152/16/128/manual_time                  -0.0034         -0.0037             2             2             2             2
Hashing/hash_partition/131072/16/256/manual_time                   -0.0542         -0.0539             0             0             0             0
Hashing/hash_partition/262144/16/256/manual_time                   -0.0472         -0.0463             0             0             0             0
Hashing/hash_partition/524288/16/256/manual_time                   -0.0382         -0.0391             1             1             1             1
Hashing/hash_partition/1048576/16/256/manual_time                  -0.0304         -0.0306             1             1             1             1
Hashing/hash_partition/2097152/16/256/manual_time                  -0.0178         -0.0187             2             2             2             2
Hashing/hash_partition/131072/16/512/manual_time                   +0.0108         +0.0198             0             0             0             0
Hashing/hash_partition/262144/16/512/manual_time                   -0.0391         -0.0391             1             0             1             1
Hashing/hash_partition/524288/16/512/manual_time                   -0.0310         -0.0311             1             1             1             1
Hashing/hash_partition/1048576/16/512/manual_time                  -0.0273         -0.0372             1             1             1             1
Hashing/hash_partition/2097152/16/512/manual_time                  -0.0043         -0.0050             2             2             2             2
Hashing/hash_partition/131072/16/1024/manual_time                  -0.0426         -0.0435             1             1             1             1
Hashing/hash_partition/262144/16/1024/manual_time                  -0.0318         -0.0319             1             1             1             1
Hashing/hash_partition/524288/16/1024/manual_time                  -0.0267         -0.0273             1             1             1             1
Hashing/hash_partition/1048576/16/1024/manual_time                 -0.0169         -0.0179             2             2             2             2
Hashing/hash_partition/2097152/16/1024/manual_time                 -0.0120         -0.0127             3             3             3             3
Hashing/hash_partition/131072/256/64/manual_time                   +0.0106         +0.0098             4             4             4             4
Hashing/hash_partition/262144/256/64/manual_time                   +0.0084         +0.0073             6             6             6             6
Hashing/hash_partition/524288/256/64/manual_time                   +0.0054         +0.0055             9             9             9             9
Hashing/hash_partition/1048576/256/64/manual_time                  +0.0160         +0.0154            17            17            17            17
Hashing/hash_partition/2097152/256/64/manual_time                  -0.0066         -0.0067            33            32            33            32
Hashing/hash_partition/131072/256/128/manual_time                  +0.0121         +0.0118             4             4             4             4
Hashing/hash_partition/262144/256/128/manual_time                  +0.0071         +0.0068             6             6             6             6
Hashing/hash_partition/524288/256/128/manual_time                  +0.0042         +0.0041             9             9             9             9
Hashing/hash_partition/1048576/256/128/manual_time                 +0.0039         +0.0039            17            17            17            17
Hashing/hash_partition/2097152/256/128/manual_time                 +0.0074         +0.0073            33            33            33            33
Hashing/hash_partition/131072/256/256/manual_time                  +0.0276         +0.0264             5             5             5             5
Hashing/hash_partition/262144/256/256/manual_time                  +0.0072         +0.0068             6             6             6             6
Hashing/hash_partition/524288/256/256/manual_time                  +0.0022         +0.0020             9             9             9             9
Hashing/hash_partition/1048576/256/256/manual_time                 -0.0035         -0.0036            17            17            17            17
Hashing/hash_partition/2097152/256/256/manual_time                 -0.0042         -0.0042            34            34            34            34
Hashing/hash_partition/131072/256/512/manual_time                  -0.0526         -0.0588             6             6             6             6
Hashing/hash_partition/262144/256/512/manual_time                  +0.0065         +0.0059             7             7             7             7
Hashing/hash_partition/524288/256/512/manual_time                  +0.0058         +0.0054            11            11            11            11
Hashing/hash_partition/1048576/256/512/manual_time                 -0.0030         -0.0030            20            20            20            20
Hashing/hash_partition/2097152/256/512/manual_time                 +0.0042         +0.0041            37            37            37            37
Hashing/hash_partition/131072/256/1024/manual_time                 +0.0075         +0.0066             7             8             7             8
Hashing/hash_partition/262144/256/1024/manual_time                 +0.0033         +0.0025             9             9             9             9
Hashing/hash_partition/524288/256/1024/manual_time                 +0.0025         +0.0019            14            14            14            14
Hashing/hash_partition/2097152/256/1024/manual_time                +0.0046         +0.0045            46            46            46            46
```

### CODE REVIEW

File: cpp/src/partitioning/partitioning.cu [hunk 1]
```diff
@@ -681,28 +676,30 @@ struct dispatch_map_type {
...
 
     // Scatter the rows into their partitions
     auto scattered =
       cudf::detail::scatter(t, scatter_map.begin(), scatter_map.end(), t, false, stream, mr);
 
-    return std::make_pair(std::move(scattered), std::move(partition_offsets));
+    // Copy offsets to host and return
+    return std::make_pair(
+      std::move(scattered),
+      std::move(partition_offsets));  // cudf::detail::make_std_vector_sync(histogram, stream));
```
Comment (vuule) [inline cpp/src/partitioning/partitioning.cu:702]: The comment is incorrect, and I'm not sure what the purpose of the commented out code is. Was there a problem with creating the vector here?

**Review by vuule — CHANGES_REQUESTED** [review]
Looks good. Just found that one place that needs to be ironed out.

File: cpp/src/partitioning/partitioning.cu (hunk 1 above)
Comment (shwina) [reply cpp/src/partitioning/partitioning.cu:702]: I never miss a chance to share this: https://nedbatchelder.com/text/deleting-code.html

Comment (harrism) [reply cpp/src/partitioning/partitioning.cu:702]: I accidentally left this in when I was debugging. Thanks for catching. I usually don't leave commented out code intentionally, and if I do I always leave a comment.

**Comment by harrism** [comment]
@vuule removed the errant comment.

**Review by vuule — APPROVED** [review]

**Comment by harrism** [comment]
rerun tests

**Review by ttnghia — APPROVED** [review]

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**